OpenRA.Mods.Common/SpriteLoaders/IcnD2Loader.cs

```diff
@@ -0,0 +1,274 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Drawing;
+using System.IO;
+using OpenRA;
+using OpenRA.FileFormats;
+using OpenRA.Graphics;
+
+namespace OpenRA.Mods.Common.SpriteLoaders
+{
+	public class IcnD2Loader : ISpriteLoader2
+	{
+		const int SizeX = 16;
+		const int SizeY = 16;
+		const int TileSize = SizeX * SizeY / 2;
+
+		ushort[][] mapTileSet;
+
+		uint ssetOffset, ssetLength;
+		uint rpalOffset, rpalLength;
+		uint rtblOffset, rtblLength;
+		uint numTiles;
+
+		byte[] rtbl;
+		byte[][] rpal;
+
+		class IcnD2Tile : ISpriteFrame
+		{
+			public Size Size { get; private set; }
+			public Size FrameSize { get { return Size; } }
+			public float2 Offset { get { return float2.Zero; } }
+			public byte[] Data { get; set; }
+			public bool DisableExportPadding { get { return false; } }
+
+			public IcnD2Tile(Stream s, byte[] pallette)
+			{
+				byte[] tile = StreamExts.ReadBytes(s, TileSize);
+
+				Size = new Size(SizeX, SizeY);
+				Data = new byte[Size.Width * Size.Height];
+
+				var i = 0;
+				for (var y = 0; y < SizeY; y++)
+					for (var x = 0; x < SizeX; x += 2)
+					{
+						byte val = tile[(y * SizeX + x) / 2];
+						Data[i++] = pallette[val >> 4];
+						Data[i++] = pallette[val & 0x0F];
+					}
+			}
+		}
+
+		class IcnD2Frame : ISpriteFrame
+		{
+			public Size Size { get; private set; }
+			public Size FrameSize { get { return Size; } }
+			public float2 Offset { get { return float2.Zero; } }
+			public byte[] Data { get; set; }
+			public bool DisableExportPadding { get { return false; } }
+
+			public IcnD2Frame(int tilesX, int tilesY, int tileIndex, IcnD2Tile[] tiles, ushort[] map)
+			{
+				Size = new Size(SizeX * tilesX, SizeY * tilesY);
+				Data = new byte[Size.Width * Size.Height];
+				var tileOffset = tilesX * tilesY;
+
+				for (var ty = 0; ty < tilesY; ty++)
+					for (var tx = 0; tx < tilesX; tx++)
+					{
+						var j = map[tileIndex * tileOffset + ty * tilesX + tx];
+						for (var y = 0; y < SizeY; y++)
+							for (var x = 0; x < SizeX; x++)
+								Data[(ty * SizeY + y) * Size.Width + (x + tx * SizeX)] = tiles[j].Data[y * SizeX + x];
+					}
+			}
+		}
+
+		bool IsIcnD2(Stream s)
+		{
+			if (s.Length < 0x20)
+				return false;
+
+			var start = s.Position;
+
+			s.Position = 0x18;
+			if (s.ReadByte() != 'S' || s.ReadByte() != 'S' || s.ReadByte() != 'E' || s.ReadByte() != 'T')
+				return false;
+
+			ssetLength = BigendianExts.ReadUInt32(s) - 8;
```

Comment:
You can use `int2.Swap(s.ReadUInt32())` in these three places, and then remove the `BigendianExts` class.